File: leetcode/algorithms/string/group-anagrams.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Group Anagrams"
*/

class Solution {
    private String getSortedString(String str){
        char[] strCharArr = str.toCharArray();
        Arrays.sort(strCharArr);
        return new String(strCharArr);
    }
    
    public List<List<String>> groupAnagrams(String[] strs) {
        
        ArrayList<List<String>> ansList = new ArrayList<List<String>>();
        HashMap<String,ArrayList<String>> stringMap = new HashMap<String,ArrayList<String>>();
        
        for(String str : strs){
            String sortedStr = getSortedString(str);
            ArrayList<String> existingList = stringMap.getOrDefault(sortedStr,null);
            if(existingList!=null){
                existingList.add(str);
            } else {
                stringMap.put(sortedStr,new ArrayList<String>(Arrays.asList(str)));
            }
        }
        
        for(Map.Entry<String,ArrayList<String>> entry : stringMap.entrySet()){
            ansList.add(entry.getValue());
        }
        
        
        return ansList;
        
        
    }
}

File: leetcode/algorithms/string/group-anagrams.ts
/*
    leetcode
    category: "Algorithms"
    problem title: "Group Anagrams"
*/

function getKey(str: string): string{
    return str.split('').sort((a,b) => a.charCodeAt(0)-b.charCodeAt(0)).join('');
}

function groupAnagrams(strs: string[]): string[][] {
    const groupsMap: Map<string,string[]> = strs.reduce((strMap,str) => {
        const key: string = getKey(str);
        strMap.set(key,(strMap.get(key) || []).concat(str));
        return strMap;
    },new Map<string,string[]>());
    
    const strGroups: string[][] = [];
    groupsMap.forEach((val,key) => {
        strGroups.push(val);
    })
    return strGroups;
};